ngx_http_sxg_filter_module.c

```diff
@@ -663,15 +710,43 @@ static bool make_chain_from_buffer(ngx_http_request_t* req,
 
 static ngx_int_t ngx_http_sxg_body_filter(ngx_http_request_t* req,
                                           ngx_chain_t* in) {
-  ngx_http_sxg_srv_conf_t* ssc =
-      ngx_http_get_module_srv_conf(req, ngx_http_sxg_filter_module);
   ngx_http_sxg_ctx_t* ctx =
       ngx_http_get_module_ctx(req, ngx_http_sxg_filter_module);
 
-  if (ctx == NULL || req->done || req->header_sent) {
+  if (ctx == NULL || ctx->finished) {
     return ngx_http_next_body_filter(req, in);
   }
 
+  if (pthread_mutex_lock(&ctx->lock) != 0) {
+    return NGX_ERROR;
+  }
+  ngx_int_t result = ngx_http_sxg_body_filter_impl(req, ctx, in);
+
+  if (result == NGX_OK || result == NGX_ERROR) {
+    ctx->finished = true;
+    if (result == NGX_ERROR) {
+      ngx_log_error(NGX_LOG_DEBUG, req->connection->log, 0,
+                    "nginx-sxg-module: finalizing SXG request context");
+    }
+    release_ctx(req->pool, ctx);
+  }
+
+  if (pthread_mutex_unlock(&ctx->lock) != 0) {
```

Comment:
This releases the context and thus destroys the mutex before unlocking it.  If you opt-into pthread error checking it should warn:

```c
    pthread_mutexattr_t attr;
    pthread_mutexattr_init(&attr);
#if !NDEBUG
    pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
#endif
    pthread_mutex_init(&ctx->lock, &attr);
```

Reply:
Moved before releasing.